openshift/odo
# url list routes and ingress

Signed-off-by: Stephanie <stephanie.cao@ibm.com>

**What type of PR is this?**
/kind bug
/area url
/area devfile

**What does does this PR do / why we need it**:
Since experimental mode now supports both ingress and route as url, we should show both types of URL created under experimental mode when running `odo url list`
Also `odo url describe` should be able to get the route URL information

```
$ odo url list
Found the following URLs for component nodejsproj
NAME            STATE               URL                                          PORT     SECURE     KIND
ingressurl2     Pushed              https://ingressurl2.1.2.3.4.com              3000     true       ingress
testroute1      Locally Deleted     http://testroute1-che.9.42.26.107.nip.io     3000     false      route
routeurl1       Pushed              http://routeurl1-che.9.42.26.107.nip.io      3000     false      route
There are local changes. Please run 'odo push'.
```

```
$ odo url describe routeurl2
NAME          STATE      URL                                          PORT     SECURE     KIND
routeurl1     Pushed     http://routeurl1-che.9.42.26.107.nip.io      3000     false       route
```

**Which issue(s) this PR fixes**:

Fixes #2985
Fixes #3304
Fixes #3274 
Fixes #3182

**How to test changes / Special notes to the reviewer**:
Automation test:
Simply run ` make test-cmd-devfile-url`

Manual test:
Enable experimental mode `odo preference set experimental true`
And create both route URL and ingress URL
```
odo url create route1
odo url create ingress1 --host 1.2.3.4.com --ingress
```

Run `url list` and `url describe` to check the route URL information is displayed correctly.
